actionpack/lib/action_dispatch/middleware/static.rb

```diff
@@ -13,11 +16,12 @@ module ActionDispatch
   # located at `public/assets/application.js` if the file exists. If the file
   # does not exist a 404 "File not Found" response will be returned.
   class FileHandler
-    def initialize(root, cache_control)
+    def initialize(root, cache_control, headers)
       @root          = root.chomp('/')
       @compiled_root = /^#{Regexp.escape(root)}/
-      headers        = cache_control && { 'Cache-Control' => cache_control }
-      @file_server = ::Rack::File.new(@root, headers)
+
+      headers ||= {}
+      @file_server = ::Rack::File.new(@root, headers.reverse_merge('Cache-Control' => cache_control))
```

Comment:
`merge` and `reverse_merge` are really expensive. Even though this only gets called once on initialization the code in this file is very performance sensitive and I don't want people copying existing style. Can you also factor out the merge?